libcudacxx/include/cuda/__mdspan/optimized_accessors.h

```diff
@@ -0,0 +1,184 @@
+
+#  include <cub/thread/thread_load.cuh>
+#  include <cub/thread/thread_store.cuh>
+
+#  include <cuda/std/__cccl/attributes.h>
+#  include <cuda/std/__cccl/compiler.h>
+#  include <cuda/std/__cccl/dialect.h>
+#  include <cuda/std/__cccl/execution_space.h>
+#  include <cuda/std/type_traits>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+enum class EvictionPolicy
+{
+  Default,
+  First,
+  Normal,
+  Last,
+  LastUse,
+  NoAllocation
+};
+
+enum class PrefetchSize
+{
+  NoPrefetch,
+  Bytes64,
+  Bytes128,
+  Bytes256
+};
+
+template <class _ElementType,
+          EvictionPolicy _Eviction = EvictionPolicy::Default,
+          PrefetchSize _Prefetch   = PrefetchSize::NoPrefetch,
+          typename _Enable         = void>
+struct cache_policy_accessor;
+
+/***********************************************************************************************************************
+ * accessor_reference
+ **********************************************************************************************************************/
+
+template <class _ElementType, EvictionPolicy _Eviction, PrefetchSize _Prefetch>
+class accessor_reference
+{
+  using pointer_type = _ElementType*;
+
+  pointer_type __p;
+
+  friend class cache_policy_accessor<_ElementType, _Eviction, _Prefetch>;
+
+public:
+  accessor_reference() = delete;
+
+  accessor_reference(accessor_reference&&) = delete;
+
+  _CCCL_HIDE_FROM_ABI _CCCL_DEVICE _CCCL_FORCEINLINE accessor_reference(const accessor_reference&) = default;
+
+  _CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE operator _ElementType() const noexcept
+  {
+    return cub::ThreadLoad<_Eviction, _Prefetch>(__p);
+  }
+
+  _CCCL_HIDE_FROM_ABI _CCCL_DEVICE _CCCL_FORCEINLINE accessor_reference& operator=(_ElementType __x) noexcept
+  {
+    return cub::ThreadStore<_Eviction>(__p, __x);
+  }
+
+  _CCCL_HIDE_FROM_ABI _CCCL_DEVICE _CCCL_FORCEINLINE accessor_reference&
+  operator=(const accessor_reference& __x) noexcept
+  {
+    return operator=(static_cast<_ElementType>(__x));
+  }
+
+private:
+  _CCCL_HIDE_FROM_ABI _CCCL_DEVICE _CCCL_FORCEINLINE explicit accessor_reference(pointer_type __p_) noexcept
+      : __p{__p_}
+  {}
+};
+
+/***********************************************************************************************************************
+ * load/store cache_policy_accessor
+ **********************************************************************************************************************/
+
+template <class _ElementType, EvictionPolicy _Eviction, PrefetchSize _Prefetch>
+struct cache_policy_accessor<_ElementType,
+                             _Eviction,
+                             _Prefetch,
+                             ::cuda::std::__enable_if_t<!::cuda::std::is_const<_ElementType>::value>>
+{
+  static_assert(!::cuda::std::is_array<_ElementType>::value,
+                "cache_policy_accessor: template argument may not be an array type");
+  static_assert(!::cuda::std::is_abstract<_ElementType>::value,
```

Comment:
Please use `_CUDA_VSTD` instead of `::cuda::std` in product code. That also includes the internal ABI namespace and makes it easier to change things when we need